### PR TITLE
`Development`: Improve client test speed and coverage for plagiarism detection

### PR DESCRIPTION
**Pull request by filip-gregurevic**
#### General

- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).

### Description

Added test to increase code coverage for plagiarism components over 80% and also refactored some tests to improve the speed.

#### Code Review
- [x] Review 1
- [x] Review 2

### Test Coverage

Bildschirmfoto 2022-01-13 um 16 53 49

### CODE REVIEW

File: src/test/javascript/spec/component/plagiarism/plagiarism-split-view.component.spec.ts [hunk 1]
```diff
@@ -128,14 +155,47 @@ describe('Plagiarism Split View Component', () => {
...
         jest.spyOn(comp, 'mapMatchesToElements').mockReturnValue(new Map());
 
-        const matches: PlagiarismMatch[] = [];
+        const matches: PlagiarismMatch[] = [
+            { startA: 0, startB: 0, length: 5 },
+            { startA: 10, startB: 10, length: 20 },
+        ];
         comp.parseTextMatches({ submissionA, submissionB, matches } as PlagiarismComparison<TextSubmissionElement>);
 
         expect(comp.matchesA).toBeDefined();
```
Comment (willergeorg) [inline src/test/javascript/spec/component/plagiarism/plagiarism-split-view.component.spec.ts:167]: Using `toBeDefined()` is violating our client test guidelines. Is it possible to specify the expectation here?

File: src/test/javascript/spec/component/plagiarism/plagiarism-split-view.component.spec.ts [hunk 2]
```diff
@@ -128,14 +155,47 @@ describe('Plagiarism Split View Component', () => {
...
         expect(comp.matchesB).toBeDefined();
```
Comment (willergeorg) [inline src/test/javascript/spec/component/plagiarism/plagiarism-split-view.component.spec.ts:168]: See my comment above

**Review by willergeorg — COMMENTED** [review]
Code changes lgtm in general. I marked you some suggestions in the code, it would be nice if you consider them. 👍

File: src/test/javascript/spec/component/plagiarism/plagiarism-split-view.component.spec.ts (hunk 1 above)
Comment (filip-gregurevic) [reply src/test/javascript/spec/component/plagiarism/plagiarism-split-view.component.spec.ts]: @willergeorg Good catch, I must have missed that. I changed it now to a test that makes more sense 👍🏼

File: src/test/javascript/spec/component/plagiarism/plagiarism-sidebar.component.spec.ts [hunk 3]
```diff
@@ -81,4 +82,15 @@ describe('Plagiarism Sidebar Component', () => {
 
         expect(comp.currentPage).toEqual(3);
     });
+
+    it('should handle changes correctly', () => {
+        comp.comparisons = [];
```
Comment (julian-christl) [inline src/test/javascript/spec/component/plagiarism/plagiarism-sidebar.component.spec.ts:87]: Not quite sure about this one. Why do you define comparisons? How does this then relate to the change and pagedComparisons?

I suggest anyway not to test with an empty array but a filled one. Two items should be enough.

File: src/test/javascript/spec/component/plagiarism/plagiarism-split-view.component.spec.ts [hunk 4]
```diff
@@ -72,14 +88,25 @@ describe('Plagiarism Split View Component', () => {
...
+        expect(comp.isModelingExercise).toBe(false);
+        expect(comp.parseTextMatches).toHaveBeenCalledTimes(1);
+    });
+
+    it('should subscribe to the split control subject', () => {
+        comp.exercise = textExercise;
+        jest.spyOn(splitControlSubject, 'subscribe');
+
+        comp.ngOnInit();
+        expect(comp.splitControlSubject.subscribe).toHaveBeenCalled();
```
Comment (julian-christl) [inline src/test/javascript/spec/component/plagiarism/plagiarism-split-view.component.spec.ts:108]:
```suggestion
        expect(comp.splitControlSubject.subscribe).toHaveBeenCalledTimes(1);
```

File: src/test/javascript/spec/component/plagiarism/plagiarism-split-view.component.spec.ts [hunk 5]
```diff
@@ -128,14 +155,59 @@ describe('Plagiarism Split View Component', () => {
...
+        const matches: PlagiarismMatch[] = [
+            { startA: 0, startB: 0, length: 5 },
+            { startA: 10, startB: 10, length: 20 },
+        ];
         comp.parseTextMatches({ submissionA, submissionB, matches } as PlagiarismComparison<TextSubmissionElement>);
 
-        expect(comp.matchesA).toBeDefined();
-        expect(comp.matchesB).toBeDefined();
         expect(comp.mapMatchesToElements).toHaveBeenCalledTimes(2);
+        expect(comp.mapMatchesToElements).toHaveBeenCalledWith(
```
Comment (julian-christl) [inline src/test/javascript/spec/component/plagiarism/plagiarism-split-view.component.spec.ts:168]: Please use `.toHaveBeenNthCalledWith` as you can pass there the index of the call.

File: src/test/javascript/spec/component/plagiarism/plagiarism-sidebar.component.spec.ts (hunk 3 above)
Comment (filip-gregurevic) [reply src/test/javascript/spec/component/plagiarism/plagiarism-sidebar.component.spec.ts]: @julian-christl Good catch! I must have forgotten to add some comparisons. I added some values now that make more sense

File: src/test/javascript/spec/component/plagiarism/plagiarism-split-view.component.spec.ts (hunk 5 above)
Comment (filip-gregurevic) [reply src/test/javascript/spec/component/plagiarism/plagiarism-split-view.component.spec.ts]: @julian-christl Thanks, I added that now.

File: src/test/javascript/spec/component/plagiarism/plagiarism-sidebar.component.spec.ts [hunk 6]
```diff
@@ -84,13 +85,30 @@ describe('Plagiarism Sidebar Component', () => {
     });
 
     it('should handle changes correctly', () => {
```
Comment (julian-christl) [inline src/test/javascript/spec/component/plagiarism/plagiarism-sidebar.component.spec.ts:87]: This tests the pagination. Please rename the test accordingly

**Review by julian-christl — CHANGES_REQUESTED** [review]

File: src/test/javascript/spec/component/plagiarism/plagiarism-sidebar.component.spec.ts (hunk 6 above)
Comment (filip-gregurevic) [reply src/test/javascript/spec/component/plagiarism/plagiarism-sidebar.component.spec.ts]: @julian-christl Good call, I renamed the test case now 👍🏼

**Review by julian-christl — APPROVED** [review]
Looks good

**Review by willergeorg — APPROVED** [review]
Changes look good to me now.

**Pull request merged**